drivers/ethernet/eth_w5500.c

```diff
@@ -0,0 +1,586 @@
+		.buf = tmp,
+		.len = ARRAY_SIZE(tmp),
+	};
+	const struct spi_buf_set rx = {
+		.buffers = &rx_buf,
+		.count = 1,
+	};
+
+	k_sem_take(&ctx->tx_rx_sem, K_FOREVER);
+	ret = spi_transceive(ctx->spi, &ctx->spi_cfg, &tx, &rx);
+	k_sem_give(&ctx->tx_rx_sem);
+
+	if (!ret) {
+		/* skip the default dummy 0x010203 */
+		memcpy(data, &tmp[3], len);
+	}
+
+	return ret;
+}
+
+static int w5500_spi_write(struct device *dev, uint32_t addr,
+			   uint8_t *data, uint32_t len)
+{
+	int ret;
+	struct w5500_runtime *ctx = dev->data;
+	uint8_t cmd[3] = {
+		addr >> 8,
+		addr,
+		W5500_SPI_WRITE_CONTROL(addr),
+	};
+	const struct spi_buf tx_buf[2] = {
+		{
+			.buf = cmd,
+			.len = ARRAY_SIZE(cmd),
+		},
+		{
+			.buf = data,
+			.len = len,
+		},
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	k_sem_take(&ctx->tx_rx_sem, K_FOREVER);
+	ret = spi_write(ctx->spi, &ctx->spi_cfg, &tx);
+	k_sem_give(&ctx->tx_rx_sem);
+
+	return ret;
+}
+
+static int w5500_readbuf(struct device *dev, uint16_t offset, uint8_t *buf,
+			 int len)
+{
+	uint32_t addr;
+	int remain = 0;
+	int ret;
+	const uint32_t mem_start = W5500_Sn_RX_MEM_START;
+	const uint16_t mem_size = W5500_RX_MEM_SIZE;
+
+	offset %= mem_size;
+	addr = mem_start + offset;
+
+	if (offset + len > mem_size) {
+		remain = (offset + len) % mem_size;
+		len = mem_size - offset;
+	}
+
+	ret = w5500_spi_read(dev, addr, buf, len);
+	if (ret || !remain) {
+		return ret;
+	}
+
+	return w5500_spi_read(dev, mem_start, buf + len, remain);
+}
+
+static int w5500_writebuf(struct device *dev, uint16_t offset, uint8_t *buf,
+			  int len)
+{
+	uint32_t addr;
+	int ret = 0;
+	int remain = 0;
+	const uint32_t mem_start = W5500_Sn_TX_MEM_START;
+	const uint32_t mem_size = W5500_TX_MEM_SIZE;
+
+	offset %= mem_size;
+	addr = mem_start + offset;
+
+	if (offset + len > mem_size) {
+		remain = (offset + len) % mem_size;
+		len = mem_size - offset;
+	}
+
+	ret = w5500_spi_write(dev, addr, buf, len);
+	if (ret || !remain) {
+		return ret;
+	}
+
+	return w5500_spi_write(dev, mem_start, buf + len, remain);
+}
+
+static int w5500_command(struct device *dev, uint8_t cmd)
+{
+	uint8_t reg;
+	uint64_t end = z_timeout_end_calc(K_MSEC(100));
+
+	w5500_spi_write(dev, W5500_S0_CR, &cmd, 1);
+	do {
+		w5500_spi_read(dev, W5500_S0_CR, &reg, 1);
+		if (end - z_tick_get() <= 0) {
+			return -EIO;
+		}
+
+		k_msleep(1);
+	} while (reg != 0);
+
+	return 0;
+}
+
+static int w5500_tx(struct device *dev, struct net_pkt *pkt)
+{
+	struct w5500_runtime *ctx = dev->data;
+	uint16_t len = net_pkt_get_len(pkt);
+	uint16_t offset;
+	uint8_t off[2];
+	int ret;
+
+	w5500_spi_read(dev, W5500_S0_TX_WR, off, 2);
+	offset = sys_get_be16(off);
+
+	if (net_pkt_read(pkt, ctx->buf, len)) {
+		return -EIO;
+	}
+
+	ret = w5500_writebuf(dev, offset, ctx->buf, len);
+	if (ret < 0) {
+		return ret;
+	}
+
+	sys_put_be16(offset + len, off);
+	ret = w5500_spi_write(dev, W5500_S0_TX_WR, off, 2);
+	w5500_command(dev, S0_CR_SEND);
```

Comment:
So you cannot tell if the packet was actually sent or not. I see in w5500_isr() that you can check on ir for S0_IR_SENDOK. Maybe there is some logic to add up here. A semaphore you'll wait (with a relevant timeout, not K_FOREVER) here, and in w5500_isr() it would give the semaphore on sendok. Something like that.

Reply:
Thanks for mentioning this. There is no standard wait time mentioned in datasheet, but 10ms seems to work with some basic tests. Fixed it.